packages/wallet/contract/src/smartWallet.js

```diff
@@ -1,90 +1,273 @@
+  assert(invitationIssuer, 'missing invitationIssuer');
   assert(storageNode, 'missing storageNode');
+  const invitationBrand = await E(invitationIssuer).getBrand();
+  const marshaller = await E(board).getReadonlyMarshaller();
 
-  const cacheStorageNode = E(storageNode).makeChildNode('cache');
+  // #region precious state
+  /**
+   * Private purses
+   *
+   * @type {MapStore<Brand, ERef<Purse<'set'>>>}
+   */
+  const brandPurses = makeScalarBigMapStore('brand purses', { durable: true });
 
-  const wallet = await makeWallet(bank, {
-    agoricNames,
-    cacheStorageNode,
-    namesByAddress,
-    // ??? why do we make this instead of passing the address itself?
-    myAddressNameAdmin,
-    zoe,
-    board,
+  /**
+   * Invitation makers yielded by offer results
+   *
+   * @type {MapStore<number, import('./continuing').InvitationMakers>}
+   */
+  const offerToInvitationMakers = makeScalarBigMapStore('invitation makers', {
+    durable: true,
   });
+  // #endregion
 
-  const admin = E(wallet).getAdminFacet();
-
-  /** @type {Record<string, ERef<Notifier<unknown>>>} */
-  const notifierParts = {
-    contacts: E(admin).getContactsNotifier(),
-    dapps: E(admin).getDappsNotifier(),
-    /** all current issuers */
-    issuers: E(admin).getIssuersNotifier(),
-    /** the latest offer or, in case of petname change, all offers */
-    offers: E(admin).getOffersNotifier(),
-    /** most recent payment */
-    payments: E(admin).getPaymentsNotifier(),
-    /** all current purses */
-    purses: E(admin).getPursesNotifier(),
-  };
-  const mutableState = fromEntries(keys(notifierParts).map(key => [key, []]));
-  const { subscriber, publisher } = makePublishKit();
-  publisher.publish({ ...mutableState });
-  entries(notifierParts).forEach(([key, notifier]) => {
-    void observeIteration(notifier, {
-      updateState: value =>
-        publisher.publish({
-          ...assign(mutableState, { [key]: value }),
-        }),
-    });
-  });
+  // #region other state management
 
-  const marshaller = wallet.getMarshaller();
+  /**
+   * What we've been told by the Bank
+   *
+   * @type {WeakMapStore<Brand, BrandDescriptor>} */
+  const brandDescriptors = makeScalarWeakMapStore();
+  // #endregion
 
+  // #region publishing
+  // NB: state size must not grow monotonically
   const myWalletStorageNode = E(storageNode).makeChildNode(address);
-  const storedSubscriber = makeStoredSubscriber(
-    subscriber,
-    myWalletStorageNode,
+
+  /** @type {PublishKit<PursesRecord>} */
+  const pursePublishKit = makePublishKit();
+  const pursesSubscriber = makeStoredSubscriber(
+    pursePublishKit.subscriber,
+    E(myWalletStorageNode).makeChildNode('purses'),
+    marshaller,
+  );
+
+  const publishPurses = async () => {
+    const entries = Array.from(brandPurses.entries());
+    // indexed the same as brand array to merge
+    const balances = await Promise.all(
+      entries.map(([_, purse]) => E(purse).getCurrentAmount()),
```

Comment:
Each purse has a balance notifier; I would expect us to use that rather than polling with `getCurrentAmount()`.